ostd/src/mm/page_table/cursor/mod.rs

```diff
@@ -126,72 +168,131 @@ impl<'rcu, M: PageTableMode, E: PageTableEntryTrait, C: PagingConstsTrait> Curso
                     let guard = unsafe { pt.make_guard_unchecked(rcu_guard) };
                     self.push_level(guard);
                     continue;
                 }
-                Child::PageTable(_) => {
-                    unreachable!();
+                ChildRef::None => None,
+                ChildRef::Frame(pa, ch_level, prop) => {
+                    debug_assert_eq!(ch_level, level);
+
+                    // SAFETY: It must be mapped into the page table.
+                    let item = unsafe { C::item_from_raw(pa, level, prop) };
+                    // Clone a copy so that the page table still has one ownership.
+                    // TODO: Provide a `PageTableItemRef` to reduce copies.
+                    let _ = ManuallyDrop::new(item.clone());
+                    Some(item)
                 }
-                Child::None => {
-                    return Ok(PageTableItem::NotMapped {
-                        va,
-                        len: page_size::<C>(level),
-                    });
+            };
+
+            return Ok((cur_va..cur_va + page_size::<C>(level), item));
+        }
+    }
+
+    /// Moves the cursor forward to the next mapped virtual address.
+    ///
+    /// If there is mapped virtual address following the current address within
+    /// next `len` bytes, it will return that mapped address. In this case, the
+    /// cursor will stop at the mapped address.
+    ///
+    /// Otherwise, it will return `None`. And the cursor may stop at any
+    /// address after `len` bytes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the length is longer than the remaining range of the cursor.
+    pub fn find_next(&mut self, len: usize) -> Option<Vaddr> {
+        self.find_next_impl(len, true, false)
+    }
+
+    /// Moves the cursor forward to the next fragment in the range.
+    ///
+    /// See [`Self::find_next`] for more details. Other than the semantics
+    /// provided by [`Self::find_next`], this method also supports finding non-
+    /// leaf entries and splitting huge pages if necessary.
+    ///
+    /// `find_leaf` specifies whether the cursor should only stop at leaf node
+    /// entries that are mapped. If not specified, it can stop at an entry at
+    /// any level, pointing to a child page table within the range.
+    ///
+    /// `split_huge` specifies whether the cursor should split huge pages when
+    /// it finds a huge page that is mapped over the required range (`len`).
+    fn find_next_impl(&mut self, len: usize, find_leaf: bool, split_huge: bool) -> Option<Vaddr> {
+        let end = self.va + len;
+        assert!(end <= self.barrier_va.end);
+
+        let rcu_guard = self.rcu_guard;
+
+        while self.va < end {
+            let cur_va = self.va;
+            let cur_va_range = self.cur_va_range();
+            let cur_entry_fits_range = cur_va == cur_va_range.start && cur_va_range.end <= end;
+            let mut cur_entry = self.cur_entry();
```

Comment:
Move `let mut cur_entry` near to `match`, which aligns with the code in `query`?